File: src/sbml/test/TestEvent.c
/**
 * \file    TestEvent.c
 * \brief   SBML Event unit tests
 * \author  Ben Bornstein
 *
 * $Id$
 * $Source$
 */
/* Copyright 2003 California Institute of Technology and
 * Japan Science and Technology Corporation.
 *
 * This library is free software; you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License as published
 * by the Free Software Foundation; either version 2.1 of the License, or
 * any later version.
 *
 * This library is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY, WITHOUT EVEN THE IMPLIED WARRANTY OF
 * MERCHANTABILITY OR FITNESS FOR A PARTICULAR PURPOSE.  The software and
 * documentation provided hereunder is on an "as is" basis, and the
 * California Institute of Technology and Japan Science and Technology
 * Corporation have no obligations to provide maintenance, support,
 * updates, enhancements or modifications.  In no event shall the
 * California Institute of Technology or the Japan Science and Technology
 * Corporation be liable to any party for direct, indirect, special,
 * incidental or consequential damages, including lost profits, arising
 * out of the use of this software and its documentation, even if the
 * California Institute of Technology and/or Japan Science and Technology
 * Corporation have been advised of the possibility of such damage.  See
 * the GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this library; if not, write to the Free Software Foundation,
 * Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA.
 *
 * The original code contained here was initially developed by:
 *
 *     Ben Bornstein
 *     The Systems Biology Markup Language Development Group
 *     ERATO Kitano Symbiotic Systems Project
 *     Control and Dynamical Systems, MC 107-81
 *     California Institute of Technology
 *     Pasadena, CA, 91125, USA
 *
 *     http://www.cds.caltech.edu/erato
 *     mailto:sbml-team@caltech.edu
 *
 * Contributor(s):
 */


#include "common/common.h"
#include "math/FormulaParser.h"

#include "SBase.h"
#include "Event.h"
#include "EventAssignment.h"
#include "Trigger.h"
#include "Delay.h"

#include <check.h>



static Event_t *E;


void
EventTest_setup (void)
{
  E = Event_create();

  if (E == NULL)
  {
    fail("Event_create() returned a NULL pointer.");
  }
}


void
EventTest_teardown (void)
{
  Event_free(E);
}


START_TEST (test_Event_create)
{
  fail_unless( SBase_getTypeCode  ((SBase_t *) E) == SBML_EVENT );
  fail_unless( SBase_getMetaId    ((SBase_t *) E) == NULL );
  fail_unless( SBase_getNotes     ((SBase_t *) E) == NULL );
  fail_unless( SBase_getAnnotation((SBase_t *) E) == NULL );

  fail_unless( Event_getId        (E) == NULL );
  fail_unless( Event_getName      (E) == NULL );
  fail_unless( Event_getTrigger   (E) == NULL );
  fail_unless( Event_getDelay     (E) == NULL );
  fail_unless( Event_getTimeUnits (E) == NULL );

  fail_unless( Event_getNumEventAssignments(E) == 0 );
}
END_TEST


START_TEST (test_Event_createWith)
{
  Event_t   *e       = Event_createWith("e1", "");


  fail_unless( SBase_getTypeCode  ((SBase_t *) e) == SBML_EVENT );
  fail_unless( SBase_getMetaId    ((SBase_t *) e) == NULL );
  fail_unless( SBase_getNotes     ((SBase_t *) e) == NULL );
  fail_unless( SBase_getAnnotation((SBase_t *) e) == NULL );

  fail_unless( Event_getName      (e) == NULL );
  fail_unless( Event_getDelay     (e) == NULL );
  fail_unless( Event_getTimeUnits (e) == NULL );

  fail_unless( Event_getNumEventAssignments(e) == 0 );

  fail_unless( !Event_isSetTrigger(e) );

  fail_unless( !strcmp(Event_getId(e), "e1") );
  fail_unless( Event_isSetId(e) );

  Event_free(e);
}
END_TEST


START_TEST (test_Event_free_NULL)
{
  Event_free(NULL);
}
END_TEST


START_TEST (test_Event_setId)
{
  char *id = "e1";


  Event_setId(E, id);

  fail_unless( !strcmp(Event_getId(E), id) );
  fail_unless( Event_isSetId(E) );

  if (Event_getId(E) == id)
  {
    fail("Event_setId(...) did not make a copy of string.");
  }

  /* Reflexive case (pathological) */
  Event_setId(E, Event_getId(E));
  fail_unless( !strcmp(Event_getId(E), id) );

  Event_setId(E, NULL);
  fail_unless( !Event_isSetId(E) );

  if (Event_getId(E) != NULL)
  {
    fail("Event_setId(E, NULL) did not clear string.");
  }
}
END_TEST


START_TEST (test_Event_setName)
{
  char *name = "Set k2 to zero when P1 <= t";


  Event_setName(E, name);

  fail_unless( !strcmp(Event_getName(E), name) );
  fail_unless( Event_isSetName(E) );

  if (Event_getName(E) == name)
  {
    fail("Event_setName(...) did not make a copy of string.");
  }

  /* Reflexive case (pathological) */
  Event_setName(E, Event_getName(E));
  fail_unless( !strcmp(Event_getName(E), name) );

  Event_setName(E, NULL);
  fail_unless( !Event_isSetName(E) );

  if (Event_getName(E) != NULL)
  {
    fail("Event_setName(E, NULL) did not clear string.");
  }
}
END_TEST


START_TEST (test_Event_setTrigger)
{
}
END_TEST


START_TEST (test_Event_setDelay)
{
}
END_TEST


START_TEST (test_Event_setTimeUnits)
{
  char *units = "second";


  Event_setTimeUnits(E, units);

  fail_unless( !strcmp(Event_getTimeUnits(E), units) );
  fail_unless( Event_isSetTimeUnits(E) );

  if (Event_getTimeUnits(E) == units)
  {
    fail("Event_setTimeUnits(...) did not make a copy of string.");
  }

  /* Reflexive case (pathological) */
  Event_setTimeUnits(E, Event_getTimeUnits(E));
  fail_unless( !strcmp(Event_getTimeUnits(E), units) );

  Event_setTimeUnits(E, NULL);
  fail_unless( !Event_isSetTimeUnits(E) );

  if (Event_getTimeUnits(E) != NULL)
  {
    fail("Event_setTimeUnits(E, NULL) did not clear string.");
  }
}
END_TEST


START_TEST (test_Event_full)
{
  ASTNode_t         *math1    = SBML_parseFormula("0");
  Trigger_t         *trigger = Trigger_createWithMath(math1);
  ASTNode_t         *math    = SBML_parseFormula("0");
  Event_t           *e       = Event_createWith("e1", "");
  EventAssignment_t *ea      = EventAssignment_createWith("k", math);

  //Event_setTrigger(e, trigger);

  Event_setName(e, "Set k2 to zero when P1 <= t");
  Event_addEventAssignment(e, ea);

  fail_unless( Event_getNumEventAssignments(e) ==  1 );
  fail_unless( Event_getEventAssignment(e, 0)  != ea );

  EventAssignment_free(ea);
  Trigger_free(trigger);
  ASTNode_free(math);
  Event_free(e);
}
END_TEST


Suite *
create_suite_Event (void)
{
  Suite *suite = suite_create("Event");
  TCase *tcase = tcase_create("Event");


  tcase_add_checked_fixture( tcase,
                             EventTest_setup,
                             EventTest_teardown );

  tcase_add_test( tcase, test_Event_create       );
  tcase_add_test( tcase, test_Event_createWith   );
  tcase_add_test( tcase, test_Event_free_NULL    );
  tcase_add_test( tcase, test_Event_setId        );
  tcase_add_test( tcase, test_Event_setName      );
  tcase_add_test( tcase, test_Event_setTrigger   );
  tcase_add_test( tcase, test_Event_setDelay     );
  tcase_add_test( tcase, test_Event_setTimeUnits );
  tcase_add_test( tcase, test_Event_full         );

  suite_add_tcase(suite, tcase);

  return suite;
}

File: src/bindings/csharp/test/sbml/TestEvent.cs
/// 
///  @file    TestEvent.cs
///  @brief   SBML Event unit tests
///  @author  Frank Bergmann (Csharp conversion)
///  @author  Akiya Jouraku (Csharp conversion)
///  @author  Ben Bornstein 
/// 
///  $Id:$
///  $HeadURL:$
/// 
///  This test file was converted from src/sbml/test/TestEvent.c
///  with the help of conversion sciprt (ctest_converter.pl).
/// 
/// <!---------------------------------------------------------------------------
///  This file is part of libSBML.  Please visit http://sbml.org for more
///  information about SBML, and the latest version of libSBML.
/// 
///  Copyright 2005-2008 California Institute of Technology.
///  Copyright 2002-2005 California Institute of Technology and
///                      Japan Science and Technology Corporation.
///  
///  This library is free software; you can redistribute it and/or modify it
///  under the terms of the GNU Lesser General Public License as published by
///  the Free Software Foundation.  A copy of the license agreement is provided
///  in the file named "LICENSE.txt" included with this software distribution
///  and also available online as http://sbml.org/software/libsbml/license.html
/// --------------------------------------------------------------------------->*/


namespace LibSBMLCSTest {

  using libsbml;

  using  System.IO;

  public class TestEvent {
    public class AssertionError : System.Exception 
    {
      public AssertionError() : base()
      {
        
      }
    }


    static void assertTrue(bool condition)
    {
      if (condition == true)
      {
        return;
      }
      throw new AssertionError();
    }

    static void assertEquals(object a, object b)
    {
      if ( (a == null) && (b == null) )
      {
        return;
      }
      else if (a.Equals(b))
      {
        return;
      }
  
      throw new AssertionError();
    }

    static void assertNotEquals(object a, object b)
    {
      if ( (a == null) && (b == null) )
      {
        throw new AssertionError();
      }
      else if (a.Equals(b))
      {
        throw new AssertionError();
      }
    }

    static void assertEquals(bool a, bool b)
    {
      if ( a == b )
      {
        return;
      }
      throw new AssertionError();
    }

    static void assertNotEquals(bool a, bool b)
    {
      if ( a != b )
      {
        return;
      }
      throw new AssertionError();
    }

    static void assertEquals(int a, int b)
    {
      if ( a == b )
      {
        return;
      }
      throw new AssertionError();
    }

    static void assertNotEquals(int a, int b)
    {
      if ( a != b )
      {
        return;
      }
      throw new AssertionError();
    }

    private Event E;

    public void setUp()
    {
      E = new  Event();
      if (E == null);
      {
      }
    }

    public void tearDown()
    {
      E = null;
    }

    public void test_Event_create()
    {
      assertTrue( E.getTypeCode() == libsbml.SBML_EVENT );
      assertTrue( E.getMetaId() == "" );
      assertTrue( E.getNotes() == null );
      assertTrue( E.getAnnotation() == null );
      assertTrue( E.getId() == "" );
      assertTrue( E.getName() == "" );
      assertEquals(E.getTrigger(),null);
      assertEquals(E.getDelay(),null);
      assertTrue( E.getTimeUnits() == "" );
      assertTrue( E.getNumEventAssignments() == 0 );
    }

    public void test_Event_createWith()
    {
      Event e = new  Event("e1", "");
      assertTrue( e.getTypeCode() == libsbml.SBML_EVENT );
      assertTrue( e.getMetaId() == "" );
      assertTrue( e.getNotes() == null );
      assertTrue( e.getAnnotation() == null );
      assertTrue( e.getName() == "" );
      assertEquals(e.getDelay(),null);
      assertTrue( e.getTimeUnits() == "" );
      assertTrue( e.getNumEventAssignments() == 0 );
      assertEquals( false, e.isSetTrigger() );
      assertTrue((  "e1" == e.getId() ));
      assertEquals( true, e.isSetId() );
      e = null;
    }

    public void test_Event_createWithLevelVersionAndNamespace()
    {
      XMLNamespaces xmlns = new  XMLNamespaces();
      xmlns.add( "http://www.sbml.org", "sbml");
      Event object1 = new  Event(2,4,xmlns);
      assertTrue( object1.getTypeCode() == libsbml.SBML_EVENT );
      assertTrue( object1.getMetaId() == "" );
      assertTrue( object1.getNotes() == null );
      assertTrue( object1.getAnnotation() == null );
      assertTrue( object1.getLevel() == 2 );
      assertTrue( object1.getVersion() == 4 );
      assertTrue( object1.getNamespaces() != null );
      assertTrue( object1.getNamespaces().getLength() == 1 );
      object1 = null;
    }

    public void test_Event_free_NULL()
    {
    }

    public void test_Event_full()
    {
      ASTNode math1 = libsbml.parseFormula("0");
      Trigger trigger = new  Trigger(math1);
      ASTNode math = libsbml.parseFormula("0");
      Event e = new  Event("e1", "");
      EventAssignment ea = new  EventAssignment("k",math);
      e.setTrigger(trigger);
      e.setName( "Set k2 to zero when P1 <= t");
      e.addEventAssignment(ea);
      assertTrue( e.getNumEventAssignments() == 1 );
      assertNotEquals(e.getEventAssignment(0),ea);
      math = null;
      e = null;
    }

    public void test_Event_setDelay()
    {
      ASTNode math1 = libsbml.parseFormula("0");
      Delay Delay = new  Delay(math1);
      E.setDelay(Delay);
      assertNotEquals(E.getDelay(),null);
      assertEquals( true, E.isSetDelay() );
      if (E.getDelay() == Delay);
      {
      }
      E.setDelay(E.getDelay());
      assertNotEquals(E.getDelay(),Delay);
      E.setDelay(null);
      assertEquals( false, E.isSetDelay() );
      if (E.getDelay() != null);
      {
      }
    }

    public void test_Event_setId()
    {
      string id =  "e1";;
      E.setId(id);
      assertTrue(( id == E.getId() ));
      assertEquals( true, E.isSetId() );
      if (E.getId() == id);
      {
      }
      E.setId(E.getId());
      assertTrue(( id == E.getId() ));
      E.setId("");
      assertEquals( false, E.isSetId() );
      if (E.getId() != null);
      {
      }
    }

    public void test_Event_setName()
    {
      string name =  "Set k2 to zero when P1 <= t";;
      E.setName(name);
      assertTrue(( name == E.getName() ));
      assertEquals( true, E.isSetName() );
      if (E.getName() == name);
      {
      }
      E.setName(E.getName());
      assertTrue(( name == E.getName() ));
      E.setName("");
      assertEquals( false, E.isSetName() );
      if (E.getName() != null);
      {
      }
    }

    public void test_Event_setTimeUnits()
    {
      string units =  "second";;
      E.setTimeUnits(units);
      assertTrue(( units == E.getTimeUnits() ));
      assertEquals( true, E.isSetTimeUnits() );
      if (E.getTimeUnits() == units);
      {
      }
      E.setTimeUnits(E.getTimeUnits());
      assertTrue(( units == E.getTimeUnits() ));
      E.setTimeUnits("");
      assertEquals( false, E.isSetTimeUnits() );
      if (E.getTimeUnits() != null);
      {
      }
    }

    public void test_Event_setTrigger()
    {
      ASTNode math1 = libsbml.parseFormula("0");
      Trigger trigger = new  Trigger(math1);
      E.setTrigger(trigger);
      assertNotEquals(E.getTrigger(),null);
      assertEquals( true, E.isSetTrigger() );
      if (E.getTrigger() == trigger);
      {
      }
      E.setTrigger(E.getTrigger());
      assertNotEquals(E.getTrigger(),trigger);
      E.setTrigger(null);
      assertEquals( false, E.isSetTrigger() );
      if (E.getTrigger() != null);
      {
      }
    }

    public void test_Event_setUseValuesFromTriggerTime()
    {
      E.setUseValuesFromTriggerTime(false);
      assertTrue( E.getUseValuesFromTriggerTime() == false );
      E.setUseValuesFromTriggerTime(true);
      assertTrue( E.getUseValuesFromTriggerTime() == true );
    }

  }
}
